rules/typescript/original.js

```diff
@@ -0,0 +1,457 @@
+
+    // Bans // @ts-<directive> comments from being used or requires descriptions
+    // after directive
+    // https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-ts-comment.md
+    // EXPLANATION: Complete banning is too strict, but requiring the
+    // description is a nice idea
+    '@typescript-eslint/ban-ts-comment': [
+      'error',
+      {
+        'ts-expect-error': 'allow-with-description',
+        'ts-ignore': 'allow-with-description',
+        'ts-nocheck': 'allow-with-description',
+        'ts-check': 'allow-with-description',
+        minimumDescriptionLength: 3,
+      },
+    ],
+
+    // Bans // tslint:<rule-flag> comments from being used.
+    // https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-tslint-comment.md
+    '@typescript-eslint/ban-tslint-comment': 'error',
+
+    // Bans specific types from being used
+    // https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md
+    // EXPLANATION: No specific types to ban
+    '@typescript-eslint/ban-types': 'off',
+
+    // Ensures that literals on classes are exposed in a consistent style
+    // https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/class-literal-property-style.md
+    '@typescript-eslint/class-literal-property-style': ['error', 'fields'],
+
+    // Enforce or disallow the use of the record type.
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-indexed-object-style.md
+    '@typescript-eslint/consistent-indexed-object-style': ['error', 'record'],
+
+    // Enforces consistent usage of type assertions
+    // https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-assertions.md
+    '@typescript-eslint/consistent-type-assertions': [
+      'error',
+      { assertionStyle: 'as', objectLiteralTypeAssertions: 'allow-as-parameter' },
+    ],
+
+    // Enforces consistent usage of type exports (regular and type-only)
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-exports.md
+    '@typescript-eslint/consistent-type-exports': ['error', { fixMixedExportsWithInlineTypeSpecifier: true }],
+
+    // Enforces consistent usage of type imports
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-imports.md
+    '@typescript-eslint/consistent-type-imports': ['error', { prefer: 'type-imports' }],
+
+    // Consistent with type definition either interface or type
+    // https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-definitions.md
+    '@typescript-eslint/consistent-type-definitions': ['error', 'type'],
```

Comment:
It is not clear how to extend global interfaces with the current configuration. The rule reports an error in the following case and honestly, I don't know how to resolve it other than disabling the rule in every such case:

```ts
declare global {
  interface HTMLElementTagNameMap {
    'vaadin-upload': Upload;
  }
}
```

<img width="944" alt="image" src="https://user-images.githubusercontent.com/5039436/169528864-bdccda2c-534e-49fb-9c4c-926a69fd7cd3.png">

Reply:
I guess the only option is to enforce `interface` instead of `type`.

Reply:
Extending the existing interface is relatively rare option for common projects. E.g., in Hilla we never adjusted the `global` module. I would suggest disabling this rule per each case with a disable comment or change this rule for the whole project.

Actually, I don't see any problem in disabling rule for specific cases with a disable comment. It is how eslint is designed to work: if you have a case where you need to disable the rule, you could simply put a disable comment there and describe why you need it. So along with the tuning eslint for your needs, you also get a documentation for unclear cases.

The only issue I could see here is that it could require some additional work because you probably have a lot of `global` extendings in the project.

Reply:
> The only issue I could see here is that it could require some additional work because you probably have a lot of global extendings in the project.

It is indeed not a good practice to have global overrides / extends, so I would avoid it as much as possible. However, considering that the config mainly targets Vaadin-based projects which are based on web components, it might be annoying to disable this rule whenever you define a web component.

Reply:
Agreed to keep this rule enabled with `type` enforced and disable it per line when necessary.